src/main/java/build/buildfarm/worker/Pipeline.java

```diff
@@ -27,22 +27,73 @@ public class Pipeline {
   private final Map<PipelineStage, Integer> stageClosePriorities;
   private Thread joiningThread = null;
   private boolean closing = false;
-  // FIXME ThreadGroup?
+  private ThreadGroup stageThreadGroup;
+  private Runnable onUncaughtFailure = null;
 
   public Pipeline() {
     stageThreads = new HashMap<>();
     stageClosePriorities = new HashMap<>();
+    stageThreadGroup =
+        new ThreadGroup("pipeline-stages") {
+          private boolean handlingUncaughtException = false;
+          private Thread callbackThread = null;
+
+          // If there is an uncaught exception in the thread group, interrupt
+          // stage threads and notify the caller to decide how to handle it
+          @Override
+          public void uncaughtException(Thread th, Throwable ex) {
+            for (Map.Entry<PipelineStage, Thread> entry : stageThreads.entrySet()) {
+              Thread stageThread = entry.getValue();
+              if (stageThread == th) {
+                PipelineStage caughtStage = entry.getKey();
+                log.log(
+                    Level.SEVERE,
+                    String.format(
+                        "%s: stage terminating due to uncaught exception", caughtStage.name()),
+                    ex);
+                break;
+              }
+            }
+            // Interrupt the entire threadgroup
+            try {
+              interrupt();
+            } catch (Exception e) {
+              log.log(Level.SEVERE, "Pipeline stage interrupt caught exception", e);
+            }
```

Comment:
There are no checked exceptions from interrupt() - you shouldn't need this block. Add a finally that sets the future if you're truly concerned about this.